[playstation] Initial contribution

FluBBaOfWard opened the pull request:
[playstation] Initial contribution

A new binding for mainly PlayStation 4 consoles.
Its main goals are to be able to change power and current game.
You can also control the PS4 UI and send/receive text from the On-Screen-Keyboard from the binding.
To read power and game you don't need any credentials but to control them you need user-credentials for your PSN user.

Here is a discussion on the forum:
https://community.openhab.org/t/playstation-binding/93789/13

J-N-K commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,821 @@
+        }
+    }
+
+    private class SocketChannelHandler extends Thread {
```
Please move this to an own class, it's too long to be an inner class. Second: Why do you need this to be a seperate thread?

FluBBaOfWard replied:
My idea was to have all the communication in it's own class, that's why it's too long. Moved out the unnecessary methods for now.

fwolter replied:
If I see correctly, you only send commands via this connection, but don't listen for data continously. So, this could be a scheduler task.

J-N-K commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,821 @@
+        try {
+            SocketChannelHandler handler = socketChannelHandler;
+            if (handler != null && isLinked(CHANNEL_2ND_SCREEN)) {
+                for (int i = 0; i < 20; i++) {
```
This may block for a long time. Wouldn't it make more sense to retry after a defined interval?

FluBBaOfWard replied:
I moved the sending to when login ok is received.

J-N-K commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,821 @@
+
+            ByteBuffer outPacket = PS4PacketHandler.makeRemoteControlPacket(PS4_KEY_OPEN_RC);
+            sendPacketEncrypted(outPacket, channel);
+            Thread.sleep(MIN_SENDKEY_DELAY);
```
You should schedule the next step instead of blocking with `Thread.sleep`

```
sendPacketEncrypted(...);
scheduler.schedule(() -> {
   <sendremotekey...>
   scheduler.schedule(() -> {
       <nextStep...>
    }, MIN_SENDKEY_DELAY, TimeUnit...);
}, ...);
```

(just to give you an idea). With JDK9 (i.e.  OH3) `CompletableFuture`s with `delayedExecutor` whpuld be a perfect match IMO. But then we will not get this merged before OH3 release

FluBBaOfWard replied:
Fixed.

J-N-K commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,821 @@
+        String applicationName = "";
+        String applicationId = "";
+
+        String[] ss = message.trim().split("\n");
```
please use meaningful names

FluBBaOfWard replied:
is rowStrings ok?

J-N-K commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,821 @@
+                    power = OnOffType.OFF;
+                }
+                if (power != null) {
+                    if (!currentPower.equals(power)) {
```
no need to check that. the framework takes care of duplicate events

FluBBaOfWard replied:
Fixed.

J-N-K commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,821 @@
+                        }
+                    }
+                    if (thing.getStatus() != ThingStatus.ONLINE) {
+                        updateStatus(ThingStatus.ONLINE);
```
just update, no need to check before

FluBBaOfWard replied:
Fixed.

J-N-K commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4PacketHandler.java:
```diff
@@ -0,0 +1,326 @@
+        return packet;
+    }
+
+    static ByteBuffer makeCommentViewrSendPacket(int i, String text) {
```
```suggestion
    static ByteBuffer makeCommentViewerSendPacket(int i, String text) {
```

FluBBaOfWard replied:
Fixed.

J-N-K commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PlayStationHandlerFactory.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @Reference(policy = ReferencePolicy.DYNAMIC)
+    protected void setLocaleProvider(LocaleProvider provider) {
```
Please use constructpr injection instead and make the `localeProvider` `final`

FluBBaOfWard replied:
Are there docs or examples of how this is done?

J-N-K replied:
remove get & set, add a constructor
```
@Activate
public PlayStationHandlerFactory(@Reference LocaleProvider provider) {
    localeProvider = provider;
}
```

and change the declarzoon to

```
private final LocaleProvider localeProvider;
```

That's all

FluBBaOfWard replied:
Thanks, fixed.

J-N-K commented on bundles/org.openhab.binding.playstation/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,61 @@
+	<config-description uri="thing-type:playstation:PS3">
+		<parameter name="ipAddress" type="text" required="true">
+			<context>network-address</context>
+			<label>IP address</label>
```
if both are allowed, better use `host` or `hostname` instead of `ipaddress`

FluBBaOfWard replied:
I changed the description to only ipAddress since I can't test with hostname in my network.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/README.md:
```diff
@@ -0,0 +1,113 @@
+
+## Thing Configuration
+
+If you want to control your PS4 the first thing you need is your user-credentials, this is a 64 characters HEX string that is easiest obtained by using PS4-waker https://github.com/dhleong/ps4-waker. The result file is called ".ps4-wake.credentials.json" in your home directory.
```
To open ports under 1000 you often need root privileges and you would only be able to do this during discovery... I might have to think this through, I want to solve it by connecting to the PSN with the help of the users PSN credentials but I haven't had enough time to do that yet.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS3Handler.java:
```diff
@@ -0,0 +1,198 @@
+    private final Logger logger = LoggerFactory.getLogger(PS3Handler.class);
+    private static final int SOCKET_TIMEOUT_SECONDS = 2;
+
+    private PS3Configuration config = getConfigAs(PS3Configuration.class);
```
Fixed

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS3Handler.java:
```diff
@@ -0,0 +1,198 @@
+
+    @Override
+    public void initialize() {
+        logger.debug("Start initializing!");
```
Fixed

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS3Handler.java:
```diff
@@ -0,0 +1,198 @@
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> timer = refreshTimer;
```
Done

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS3Handler.java:
```diff
@@ -0,0 +1,198 @@
+     * Sets up a timer for querying the PS3 (using the scheduler) every 10 seconds.
+     */
+    private void setupRefreshTimer() {
+        ScheduledFuture<?> timer = refreshTimer;
```
Fixed

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Command.java:
```diff
@@ -0,0 +1,85 @@
+    COMMENT_VIEWER_SEND(0x32),
+    HELLO_REQ(0x6f636370);
+
+    private static final Map<Integer, PS4Command> TAG_MAP = new HashMap<>();
```
Can't get it to work, says "Cannot invoke stream() on the array type PS4Command[]". Do you have an example (plug-in) where this is used?

J-N-K replied:
`Arrays.stream(PS4Command.values()).collect(...)`

FluBBaOfWard replied:
Thanks. Fixed now.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS3Handler.java:
```diff
@@ -0,0 +1,198 @@
+    private @Nullable ScheduledFuture<?> refreshTimer;
+
+    // State of PS3
+    private OnOffType currentPower = OnOffType.OFF;
```
Fixed

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS3Handler.java:
```diff
@@ -0,0 +1,198 @@
+
+    private void turnOnPS3() {
+
+        try (DatagramSocket srchSocket = new DatagramSocket(); DatagramSocket wakeSocket = new DatagramSocket();) {
```
The packet sent to the PS3 is "SRCH", I changed the name of the socket to searchSocket.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS3Handler.java:
```diff
@@ -0,0 +1,198 @@
+            // wait for responses
+            byte[] rxbuf = new byte[256];
+            DatagramPacket receivePacket = new DatagramPacket(rxbuf, rxbuf.length);
+            for (int i = 0; i < 34; i++) {
```
Uhm, yes. I guess I need another approach?

J-N-K replied:
Why not extract the send/check to another method, scheduler.execute(---) that method and re-schedule in that mmethod if connection attempt fails?

FluBBaOfWard replied:
Done.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS3Handler.java:
```diff
@@ -0,0 +1,198 @@
+        for (int i = 0; i < 6; i++) {
+            packet[pos++] = -1; // 0xFF
+        }
+        for (int j = 0; j < 16; j++) {
```
Fixed.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,821 @@
+
+    @Override
+    public void handleConfigurationUpdate(Map<String, Object> configurationParameters) {
+        super.handleConfigurationUpdate(configurationParameters);
```
So initialize is run every time the configuration is changed? Then I can move the code there.

FluBBaOfWard replied:
I switched the order of the arguments so that login() is only called when the user have entered a pairing code.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Configuration.java:
```diff
@@ -0,0 +1,98 @@
+     * Constant field used in {@link PlayStationDiscovery} to set the configuration property during discovery.
+     * Value of this field needs to match {@link #userCredential}.
+     */
+    public static final String USER_CREDENTIAL = "userCredential";
```
Fixed.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4ArtworkHandler.java:
```diff
@@ -0,0 +1,139 @@
+     * @return A JPEG image as a RawType if an art work file is found otherwise null.
+     */
+    @Nullable
+    RawType fetchArtworkForTitleid(String titleid, Integer size, Locale locale) {
```
Is package private not good in some way?

FluBBaOfWard replied:
Set it to public.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4ArtworkHandler.java:
```diff
@@ -0,0 +1,139 @@
+    static final String SERVICE_CATEGORY = "binding";
+
+    /** Service pid */
+    static final String SERVICE_PID = "org.openhab." + SERVICE_CATEGORY + "." + SERVICE_ID;
```
Looked at some example... fixed now.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4ArtworkHandler.java:
```diff
@@ -0,0 +1,139 @@
+    /** Cache folder under $userdata */
+    private static final String CACHE_FOLDER_NAME = "cache";
+
+    PS4ArtworkHandler() {
```
Is there some policy/common understanding (OpenHab or other) against package private? I'm still learning Java so I have no strong beliefs either way, but a why goes a long way.

cpmeister replied:
Just speaking from personal opinion here as seasoned java developer. But package private is kinda the ugly duckling of access modifiers. Package access isn't used very often because anytime you want to set access between `private` and `public` the goto is `protected` since `protected` includes package protection as well as give subclasses access. Since Java is an object-oriented programming (OOP) language, packages are generally seen as a way to organize code, not as a way to architecturally structure code. Package private is an architectural decision whereas `protected` fits into OOP design.

Personally, I wish `protected` didn't include package private access since it is rarely my intention to give other package classes access to protected fields.

I don't know if my explanation makes any sense, it is hard to put into words.

FluBBaOfWard replied:
The constructor is now private since the whole class is now static.

FluBBaOfWard commented:
I'm not sure what I need to do right now to move on...

fwolter commented:
What's the state of this PR? Are you still interested in a review?

FluBBaOfWard commented:
Yes, please.

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,845 @@
+                    }
+                }
+            } catch (IOException e) {
+                logger.info("Connection read exception: {}", e.getMessage());
```
See above. Logging to info. Same for below.

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,845 @@
+                                    ByteBuffer outPacket = PS4PacketHandler
+                                            .makeClientIDPacket("com.playstation.mobile2ndscreen", "18.9.3");
+                                    sendPacketEncrypted(outPacket, false);
+                                }, 10, TimeUnit.MILLISECONDS);
```
10ms can be executed in the same time slice, as it is too short for the JVM. If you need a minimal delay, 100ms would be better.

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,845 @@
+            }
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            logger.info("Send packet {} exception: {}", cmd, e.getMessage());
```
See above

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,845 @@
+            }
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            logger.info("Send packet exception: {}", e.getMessage());
```
See above

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,845 @@
+            login(channel);
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            logger.info("Send packet exception: {}", e.getMessage());
```
See above

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,845 @@
+
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            logger.info("RemoteKey exception: {}", e.getMessage());
```
See above

FluBBaOfWard replied:
Missed this one, will fix next commit.

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4PacketHandler.java:
```diff
@@ -0,0 +1,325 @@
+@NonNullByDefault
+public class PS4PacketHandler {
+
+    private static final String APPLICATION_NAME = "OpenHAB PlayStation 4 Binding";
+    private static final String DEVICE_NAME = "OpenHAB Server";
```
```suggestion
    private static final String APPLICATION_NAME = "openHAB PlayStation 4 Binding";
    private static final String DEVICE_NAME = "openHAB Server";
```

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/discovery/PlayStationDiscovery.java:
```diff
@@ -0,0 +1,548 @@
+            NetworkInterface nic = NetworkInterface.getByInetAddress(localAddress);
+            byte[] macAdr = nic.getHardwareAddress();
+            StringBuilder macBuilder = new StringBuilder();
+            for (int macIndex = 0; macIndex < macAdr.length; macIndex++) {
+                macBuilder.append(String.format("%02x", macAdr[macIndex]));
```
You could utilize `HexUtils` here, too.

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/discovery/PlayStationDiscovery.java:
```diff
@@ -0,0 +1,548 @@
+        }
+    }
+
+    /**
+     * This method is used to find all PS3 devices that are in Remote-Play mode.
+     * Unused as of now.
+     */
+    @SuppressWarnings("unused")
+    private synchronized void discoverPS3Old() {
```
Actually, dead code shouldn't be part of a binding. Can you save it and remove it for now?

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/discovery/PlayStationDiscovery.java:
```diff
@@ -0,0 +1,548 @@
+        ThingUID uid = hostType.equalsIgnoreCase("PS5") ? new ThingUID(THING_TYPE_PS5, hostId)
+                : new ThingUID(THING_TYPE_PS4, hostId);
+
+        DiscoveryResult result = DiscoveryResultBuilder.create(uid).withProperties(properties).withLabel(hostName)
+                .build();
```
Can you add a representation property? Same for below

fwolter commented on bundles/org.openhab.binding.playstation/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,69 @@
+		</parameter>
+		<parameter name="pairingCode" type="text" min="8" max="8" pattern="[0-9]*">
+			<label>Pairing code</label>
+			<description>Code to pair OpenHAB device to PlayStation 4. Only needed during pairing.</description>
```
```suggestion
			<description>Code to pair openHAB device to PlayStation 4. Only needed during pairing.</description>
```

fwolter commented on bundles/org.openhab.binding.playstation/src/main/resources/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,69 @@
+		</parameter>
+	</config-description>
+
+	<config-description uri="thing-type:playstation:PS4">
```
Can you add these config parameters to the readme?

fwolter replied:
Did you push your changes?

FluBBaOfWard replied:
I added all configs to the demo thing example, should there be a more expansive description for all config items in the readme?

fwolter replied:
Yes, there should be a table like here https://www.openhab.org/addons/bindings/telegram/#thing-configuration

fwolter commented on bundles/org.openhab.binding.playstation/src/main/resources/ESH-INF/i18n/playstation_en.properties:
```diff
@@ -0,0 +1,66 @@
+thing-type.config.playstation.PS4.passCode.label = Pass code
+thing-type.config.playstation.PS4.passCode.description = Pass code to log in to PlayStation 4, optional, 4 digits.
+thing-type.config.playstation.PS4.pairingCode.label = Pairing code
+thing-type.config.playstation.PS4.pairingCode.description = Pairing code to pair OpenHAB device to PlayStation 4, only needed during pairing, 8 digits.
```
```suggestion
thing-type.config.playstation.PS4.pairingCode.description = Pairing code to pair openHAB device to PlayStation 4, only needed during pairing, 8 digits.
```

fwolter reviewed (changes requested):
I reviewed your code and here is my feedback.

FluBBaOfWard commented:
I hope I have fixed everything except changing some of the Switches to Actions, will fix that during the weekend.
Should I press the resolve button for things I have fixed or do you do that when you are satisfied with the result?

fwolter commented:
You can press resolve to show, which comments you already fixed.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4ArtworkHandler.java:
```diff
@@ -0,0 +1,160 @@
+            }
+        }
+        String request = buildArtworkRequest(locale, titleId, size);
+        artwork = HttpUtil.downloadImage(request, 1000);
```
I used the default impl without specified time instead.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,845 @@
+                    case CHANNEL_LOG_OUT:
+                        logOut();
+                        break;
+                    case CHANNEL_SCREEN_SHOT:
```
LogOut and ScreenShot removed since I didn't get them to do anything on the PS4.

FluBBaOfWard commented:
The SocketChannelHandler class handles all the responses from the PS4.

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4ArtworkHandler.java:
```diff
@@ -0,0 +1,160 @@
+            } catch (FileNotFoundException ex) {
+                LOGGER.debug("Could not find {} in cache. {}", artworkFileInCache, ex.getMessage());
+            } catch (IOException ex) {
+                LOGGER.info("Could not read {} from cache. {}", artworkFileInCache, ex.getMessage());
```
Logging to info should be used rarely. This could be debug. Same for below.

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java:
```diff
@@ -0,0 +1,837 @@
+                sendPacketToPS4(packet, channel, true, requiresLogin);
+            }
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            logger.debug("Send packet exception: {}", e.getMessage());
```
The log message could be removed, as the status update is already logged by the framework. Same for below.

fwolter commented on bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/discovery/PlayStationDiscovery.java:
```diff
@@ -0,0 +1,508 @@
+        ThingUID uid = hostType.equalsIgnoreCase("PS5") ? new ThingUID(THING_TYPE_PS5, hostId)
+                : new ThingUID(THING_TYPE_PS4, hostId);
+
+        DiscoveryResult result = DiscoveryResultBuilder.create(uid).withProperties(properties).withLabel(hostName)
+                .withRepresentationProperty(hostId).build();
```
The representation property must be the *name* of the property, not the value. Same for below.

fwolter commented on bundles/org.openhab.binding.playstation/README.md:
````diff
@@ -0,0 +1,118 @@
+}
+```
+
+## Any custom content here!
````
I would call this "Caveat and Limitations".

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/README.md:
```diff
@@ -0,0 +1,117 @@
+| applicationImage| Image     | Application artwork.                                                     | R          |
+| oskText         | String    | The text from the OnScreenKeyboard.                                      | RW         |
+| disconnect      | Switch    | Disconnect from PS4.                                                     | W          |
+| keyUp           | Switch    | Push Up button.                                                          | W          |
+| keyDown         | Switch    | Push Down button.                                                        | W          |
+| keyRight        | Switch    | Push Right button.                                                       | W          |
+| keyLeft         | Switch    | Push Left button.                                                        | W          |
+| keyEnter        | String    | Push Enter button.                                                       | W          |
+| keyBack         | Switch    | Push Back button.                                                        | W          |
+| keyOption       | Switch    | Push Option button.                                                      | W          |
+| keyPS           | Switch    | Push PS button.                                                          | W          |
```
It feels like Actions are way too advanced to do something so simple as to only send a message. Are triggers only for channels that emit an event? But if that is the way they should be I will fix it but it will take some more time then a weekend.

fwolter replied:
You could also consolidate these Channels into one Channel with a Command Description. This should be way easier as migrating to Actions. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#command-description

FluBBaOfWard replied:
Thanks, will definitely try that instead.

FluBBaOfWard commented on bundles/org.openhab.binding.playstation/README.md:
```diff
@@ -0,0 +1,117 @@
+| applicationId   | String    | Id of the currently running application.                                 | RW         |
+| applicationImage| Image     | Application artwork.                                                     | R          |
+| oskText         | String    | The text from the OnScreenKeyboard.                                      | RW         |
+| disconnect      | Switch    | Disconnect from PS4.                                                     | W          |
```
I changed this to "Connect" instead and made it work both ways, so now you can connect without sending a command and you can also read the current status.

fwolter replied:
When does the user need to connect or disconnect from the PS4?

FluBBaOfWard commented:
I made a new PR for openHAB 3

fwolter commented:
Closed in favor of #8585
